### ISSUE TITLE
Dart compiler exited unexpectedly after flutter upgrade

### ISSUE THREAD

**Issue by gyuson** [flutter/flutter#148611]
### Steps to reproduce

1. Upgraded Flutter (3.22.0) and Dart (3.4.0)
2. updated dependences 

Ran into the following errors repeatedly, even after cleaning, fetching dependencies again, etc.

```
`Launching lib/main.dart on Pixel 7a in debug mode...
Unhandled exception:
Crash when compiling:
Concurrent modification during iteration: Instance(length:2) of '_GrowableList'.

#0      ListIterator.moveNext (dart:_internal/iterable.dart:348)
#1      BodyBuilder._unaliasTypeAliasedConstructorInvocations
(package:front_end/src/fasta/kernel/body_builder.dart:1640)
#2      BodyBuilder.performDelayedActions
(package:front_end/src/fasta/kernel/body_builder.dart:1755)
#3      SourceLoader.buildOutlineExpressions
(package:front_end/src/fasta/source/source_loader.dart:2820)
#4      KernelTarget.buildOutlines.<anonymous closure>
(package:front_end/src/fasta/kernel/kernel_target.dart:545)
<asynchronous suspension>
#5      withCrashReporting (package:front_end/src/fasta/crash.dart:133)
<asynchronous suspension>
#6      KernelTarget.buildOutlines (package:front_end/src/fasta/kernel/kernel_target.dart:415)
<asynchronous suspension>
#7      IncrementalCompiler.computeDelta.<anonymous closure>
(package:front_end/src/fasta/incremental_compiler.dart:367)
<asynchronous suspension>
#8      CompilerContext.clear (package:front_end/src/fasta/compiler_context.dart:139)
<asynchronous suspension>
#9      IncrementalCompiler.compile (package:vm/incremental_compiler.dart:77)
<asynchronous suspension>
#10     FrontendCompiler.compile (package:frontend_server/frontend_server.dart:617)
<asynchronous suspension>
#11     listenAndCompile.<anonymous closure>
(package:frontend_server/frontend_server.dart:1311)
<asynchronous suspension>
the Dart compiler exited unexpectedly.
the Dart compiler exited unexpectedly.`
```

### Expected results

Expected the app to run on debug smoothly, both in the simulator and physical device. I only had the issue after I updated the SDKs. 

### Actual results

App not running as expected

### Code sample

<details open><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

**Comment by gyuson** [comment]
flutter doctor results:
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.22.0, on macOS 14.4.1 23E224 darwin-arm64, locale en-SG)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)
[✓] VS Code (version 1.89.1)
[✓] Connected device (6 available)
    ! Error: Browsing on the local area network for Ghia’s Mini. Ensure the device is unlocked and attached with a cable or
      associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
[✓] Network resources

• No issues found!

**Comment by justprodev** [comment]
```Launching lib/main.dart on iPhone 12 in debug mode...
Unhandled exception:
Crash when compiling:
Concurrent modification during iteration: Instance(length:2) of '_GrowableList'.

#0      ListIterator.moveNext (dart:_internal/iterable.dart:348)
#1      BodyBuilder._unaliasTypeAliasedConstructorInvocations (package:front_end/src/fasta/kernel/body_builder.dart:1640)
#2      BodyBuilder.performDelayedActions (package:front_end/src/fasta/kernel/body_builder.dart:1755)
#3      SourceLoader.buildOutlineExpressions (package:front_end/src/fasta/source/source_loader.dart:2820)
#4      KernelTarget.buildOutlines.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:545)
<asynchronous suspension>
#5      withCrashReporting (package:front_end/src/fasta/crash.dart:133)
<asynchronous suspension>```

**Comment by darshankawar** [comment]
> 2\. updated dependences

@gyuson Can you provide pubspec.yaml in which you updated dependencies ?

A similar issue was fixed last year https://github.com/flutter/flutter/issues/131415

Meanwhile, tagging @mraleph for thoughts on the reported error

**Comment by gyuson** [comment]
@darshankawar here is the pubspec.yaml

name: ##########
description: "##########"

# Prevent accidental publishing to pub.dev.
publish_to: 'none'

version: 1.0.0+1

environment:
  sdk: '>=3.2.3 <4.0.0'

dependencies:
  flutter:
    sdk: flutter
  firebase_auth: ^4.17.4
  firebase_core: ^2.24.2
  get: ^4.6.6
  get_storage: ^2.1.1
  video_player: ^2.8.2
  cloud_firestore: ^4.15.4
  firebase_storage: ^11.6.5
  google_sign_in: ^6.2.1
  sign_in_with_apple: ^6.1.0
  crypto: ^3.0.3
  flutter_facebook_auth: ^6.0.4
  intl: ^0.19.0
  flutter_spinkit: ^5.2.0
  flutter_svg: ^2.0.9
  carousel_slider: ^4.2.1
  sleek_circular_slider: ^2.0.1
  screenshot: ^2.5.0
  pdf: ^3.10.7
  share_plus: ^9.0.0
  path_provider: ^2.1.2
  flutter_screenutil: ^5.9.0
  snapping_bottom_sheet: ^1.0.4
  web: ^0.5.1
  shimmer: ^3.0.0
  cached_network_image: ^3.3.1
  infinite_carousel: ^1.1.1
  flutter_local_notifications: ^17.0.0
  timezone: ^0.9.2
  webview_flutter: ^4.7.0
  json_annotation: ^4.9.0

dev_dependencies:
  flutter_test:
    sdk: flutter
  mockito: ^5.4.4
  firebase_auth_mocks: ^0.13.0 
  google_sign_in_mocks: ^0.3.0 
  fake_cloud_firestore: ^2.4.9 
  build_runner: ^2.1.4 
  flutter_lints: ^4.0.0
  json_serializable: ^6.8.0
  integration_test:
    sdk: flutter

flutter:
  fonts:
    - family: GeneralSans
      fonts:
        - asset: assets/fonts/GeneralSans-Regular.otf
        - asset: assets/fonts/GeneralSans-Medium.otf
          weight: 500
        - asset: assets/fonts/GeneralSans-Semibold.otf
          weight: 600
    - family: PlayfairDisplay
      fonts:
       - asset: assets/fonts/PlayfairDisplay-Medium.ttf
       - asset: assets/fonts/PlayfairDisplay-MediumItalic.ttf
         style: italic

  uses-material-design: true

  # Enable generation of localized Strings from arb files.
  generate: true

  assets:
    - assets/images/
    - assets/images/icons/
    - assets/videos/

**Comment by justprodev** [comment]
TL;DR: In my case, the reason was that I was using deprecated MaterialState* classes. I replaced them with WidgetState and the problem went away.

But, to see the cause of this I applied dirty patch to dart-sdk (3.4.0):

<details>
  <summary>patched body_builder.dart</summary>
  
```
diff --git a/pkg/front_end/lib/src/fasta/kernel/body_builder.dart b/pkg/front_end/lib/src/fasta/kernel/body_builder.dart
index cc3d5f29e9e..b779a3137f3 100644
--- a/pkg/front_end/lib/src/fasta/kernel/body_builder.dart
+++ b/pkg/front_end/lib/src/fasta/kernel/body_builder.dart
@@ -1636,6 +1636,7 @@ class BodyBuilder extends StackListenerImpl
   void _unaliasTypeAliasedConstructorInvocations(
       List<TypeAliasedConstructorInvocation>
           typeAliasedConstructorInvocations) {
+    List<TypeAliasedConstructorInvocation> newInvocations = [];
     for (TypeAliasedConstructorInvocation invocation
         in typeAliasedConstructorInvocations) {
       assert(invocation.hasBeenInferred || isOrphaned(invocation),
@@ -1664,12 +1665,13 @@ class BodyBuilder extends StackListenerImpl
             isConst: invocation.isConst);
       }
       if (replacement == null) {
-        delayedTypeAliasedConstructorInvocations.add(invocation);
+        newInvocations.add(invocation);
       } else {
         invocation.parent?.replaceChild(invocation, replacement);
       }
     }
     typeAliasedConstructorInvocations.clear();
+    delayedTypeAliasedConstructorInvocations.addAll(newInvocations);
   }
 
   void _unaliasTypeAliasedFactoryInvocations(
@@ -1757,6 +1759,7 @@ class BodyBuilder extends StackListenerImpl
       if (delayedTypeAliasedConstructorInvocations.isNotEmpty) {
         for (ConstructorInvocation invocation
             in delayedTypeAliasedConstructorInvocations) {
+              print(invocation);
           internalProblem(
               fasta.templateInternalProblemUnhandled.withArguments(
                   invocation.target.name.text, 'performDelayedActions'),
```
  
</details>

Anyway, I think that Concurrent modification error should be avoided, even in cases like that.

Well, it would be great to show some kind of informative message.

**Comment by darshankawar** [comment]
Thanks for the update. I'll keep the issue open on the reported behavior for team's attention.

**Comment by mraleph** [comment]
@johnniwinther seems like a CFE bug. I don't know if it is already fixed upstream or not.

**Comment by johnniwinther** [comment]
It looks like a CFE bug.

**Comment by gyuson** [comment]
guys, thanks for picking this up. if you need any more details or specific information from my end to assist with the investigation or to reproduce the issue, please let me know. thanks!

**Comment by chloestefantsova** [comment]
I was able to reproduce the issue and have just landed the fix: https://github.com/dart-lang/sdk/commit/991fbf4acaca1338662583abe7a3f5bc9a3e2392.

**Comment by a-siva** [comment]
Can we close this issue as a fix has landed.

**Comment by darshankawar** [comment]
Closing as fixed.

**Comment by mraleph** [comment]
@iostriz I don't think the fix was cherry picked into 3.22.2.
